src/main.cpp

```diff
@@ -859,22 +859,39 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
         CTxMemPoolEntry entry(tx, nFees, GetTime(), dPriority, chainActive.Height(), mempool.HasNoInputsOf(tx));
         unsigned int nSize = entry.GetTxSize();
 
+        // Try to make space in mempool
+        std::set<uint256> protect;
+        BOOST_FOREACH(const CTxIn& in, tx.vin) {
+            protect.insert(in.prevout.hash);
+        }
+        std::set<uint256> stagedelete;
+        CAmount nFeesDeleted = 0;
+        size_t nSizeDeleted = 0;
+        if (!mempool.StageTrimToSize(GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000, nFees, entry.GetFeeRate(), protect, stagedelete, nFeesDeleted, nSizeDeleted)) {
+            return state.DoS(0, false, REJECT_INSUFFICIENTFEE, "mempool-full");
+        }
+
+        if ((double)nFeesDeleted * nSize > (double)nFees * nSizeDeleted) {
+            // The new transaction's feerate is below that of the replaced transactions.
+            return state.DoS(0, false, REJECT_INSUFFICIENTFEE, "mempool-full-no-improve");
+        }
+
         // Don't accept it if it can't get into a block
         CAmount txMinFee = GetMinRelayFee(tx, nSize, true);
-        if (fLimitFree && nFees < txMinFee)
+        if (fLimitFree && nFees - nFeesDeleted < txMinFee)
             return state.DoS(0, error("AcceptToMemoryPool: not enough fees %s, %d < %d",
```

Comment:
You don't want to use modified fees for the "absurdly large fees" check, I think.